Added logger to UseWebjobsStartup and UseExternalStartup

madelinegordon opened the pull request.

madelinegordon committed:
services diff comparison

brettsam commented on src/Microsoft.Azure.WebJobs.Host/Hosting/ServiceCollectionDecorator.cs:
```diff
@@ -0,0 +1,98 @@
+
+namespace Microsoft.Azure.WebJobs.Host.Hosting
+{
+    public class ServiceCollectionDecorator : IServiceCollection, ITrackingServiceCollection
```
Make this `internal`.

fabiocav replied:
Also change the type name to `TrackedServiceCollection`

brettsam commented on src/Microsoft.Azure.WebJobs.Host/Hosting/ITrackingServiceCollection.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace Microsoft.Azure.WebJobs.Host.Hosting
+{
+    interface ITrackingServiceCollection
```
Explicitly make this `internal`.

brettsam commented on src/Microsoft.Azure.WebJobs.Host/Hosting/ITrackingServiceCollection.cs:
```diff
@@ -0,0 +1,18 @@
+    {
+        IServiceCollection ServiceCollection { get; set; }
+
+        List<ServiceDescriptor> Diff { get; }
```
It looks like this is used in an iterator only, so make this `IEnumerable<ServiceDescriptor>`.

brettsam replied:
Let's call this something other than `Diff` -- maybe `ExternalServices`?

fabiocav replied:
`TrackedServices` or `TrackedChanges` would be my recommendation

madelinegordon replied:
> It looks like this is used in an iterator only, so make this `IEnumerable<ServiceDescriptor>`.

In ServiceCollectionDecorator, Add, Remove, and Clear are called on the object. Is this suggestion to make this an IEnumerable here in the interface and then cast to a List in the decorator?

brettsam commented on src/Microsoft.Azure.WebJobs.Host/Hosting/ServiceCollectionDecorator.cs:
```diff
@@ -0,0 +1,98 @@
+            Diff = new List<ServiceDescriptor>();
+        }
+
+        public ServiceDescriptor this[int index] { get => ServiceCollection[index]; set => throw new NotImplementedException(); }
```
I think the setter here should do `ServiceCollection[index] = value`

brettsam commented on src/Microsoft.Azure.WebJobs.Host/Hosting/ServiceCollectionDecorator.cs:
```diff
@@ -0,0 +1,98 @@
+            if (!Diff.Any(x => x.ServiceType == item.ServiceType)) {
+                Diff.Add(item);
+            }
+        }
```
We probably shouldn't filter here -- it's legal to register multiple `ServiceDescriptor`s for the same `ServiceType`. The last one will "win", and it'd be good to have the true representation. Unless this hit a snag somewhere...

fabiocav replied:
👍

brettsam commented on src/Microsoft.Azure.WebJobs.Host/Hosting/WebJobsBuilder.cs:
```diff
@@ -12,9 +13,10 @@ internal class WebJobsBuilder : IWebJobsBuilder
         {
-            Services = services ?? throw new ArgumentNullException(nameof(services));
+            if (services == null) { throw new ArgumentNullException(nameof(services)); } 
+            Services = new ServiceCollectionDecorator(services);
```
nit -- can you do `Services = new ServiceCollectionDecorator(services) ?? throw new ArgumentNullException(nameof(services));`

fabiocav replied:
Not sure the above would work... the `null` coalescing operator here wouldn't evaluate to `null` here (you'd just end up with the null exception from the constructor when it validates its arguments)

fabiocav replied:
my nit here would be to match the following for consistency:
``` csharp
if (services == null) 
{ 
    throw new ArgumentNullException(nameof(services)); 
} 
```

madelinegordon replied:
> my nit here would be to match the following for consistency:
> 
> ```cs
> if (services == null) 
> { 
>     throw new ArgumentNullException(nameof(services)); 
> } 
> ```
I am not sure I understand fully--is this a formatting nit? Or something else that I am missing?

brettsam replied:
Oh right, duh. 

@madelinegordon -- yeah, it's formatting, we put curly braces on newlines by themselves. I think we even have a build step that will log warnings if you don't do that (but maybe not in this repo?)

madelinegordon replied:
Ooh okay, good to know. Thank you!

brettsam commented on src/Microsoft.Azure.WebJobs.Host/Hosting/WebJobsBuilderExtensions.cs:
```diff
@@ -58,22 +61,72 @@ public static IWebJobsBuilder UseHostId(this IWebJobsBuilder builder, string hos
+                    }
+                 
+                    string lifetime = service.Lifetime.ToString(); 
+                    logger.LogInformation("Service type: {serviceName}, Implementation: {implementation}, Registered with lifetime: {lifetime}", serviceName, implementation, lifetime);
```
I would make this `LogDebug`. We'll see it in Kusto then, but customers won't see it by default. I've seen some people that register dozens of services, so logging as Info could be overload.

brettsam replied:
Oh another thing here -- we should log this with an EventId as that greatly improves querying in kusto. You should just be able to use the LogInformation overload that takes an event. and make it something like `new EventId(500, "ExternalStartupServices")`

brettsam commented on src/Microsoft.Azure.WebJobs.Host/Hosting/WebJobsBuilderExtensions.cs:
```diff
@@ -58,22 +61,72 @@ public static IWebJobsBuilder UseHostId(this IWebJobsBuilder builder, string hos
+                    }
+                 
+                    string lifetime = service.Lifetime.ToString(); 
+                    logger.LogInformation("Service type: {serviceName}, Implementation: {implementation}, Registered with lifetime: {lifetime}", serviceName, implementation, lifetime);
```
Couple of suggestions on the message:
1. Instead of `Implementation: {implementation}`, you can do something like `{registeredServiceString}` and when you build it up, make the whole string be `Instance: {instanceType}` or `Implementation: {implemenationType}` or `Factory`.
2. My opinion -- I think we should compact this and format all of them into a single message using a `StringBuilder`. It'll be easier to find in the logs.

Something like: 
```
Services registered by external startup:
  {serviceName}, {registeredServiceString}, {lifetime}
  {serviceName}, {registeredServiceString}, {lifetime}
  {serviceName}, {registeredServiceString}, {lifetime}
```

fabiocav replied:
I like the suggestion. The one thing I'd add is the external startup type name to the header.

brettsam commented on src/Microsoft.Azure.WebJobs.Host/Hosting/WebJobsBuilderExtensions.cs:
```diff
@@ -58,22 +61,72 @@ public static IWebJobsBuilder UseHostId(this IWebJobsBuilder builder, string hos
             return builder;
         }
 
+        private static void ConfigureAndLogUserConfiguredServices(IWebJobsStartup startup, IWebJobsBuilder builder, ILogger logger)
```
Rather than receive an `ILogger`, I think we should take an `ILoggerFactory` (there's a `NullLoggerFactory.Instance` as well). That way, we control what the category of that logger is.

Then down here we can call `loggerFactory.CreateLogger<ServiceCollectionDecorator>` and we'll then be in control of exactly what the category is.

madelinegordon replied:
That makes sense--I have one question though. If the logger is created within the scope of UseWebjobsStartup, how will I be able to access it in order to test that the expected services were logged? Is there an accessor to a factory's loggers?

brettsam replied:
We have a `TestLoggerProvider` in the test code. So you create a `LoggerFactory`, add the test logger provider, then pass that `LoggerFactory` into your test. That will track Loggers created, all log messages, etc. If you search in the test code for where we use `TestLoggerProvider` you should find some examples.

madelinegordon replied:
Ooh okay, I had tried that but registered a logger with the provider instead of using provider.GetAllLogMessages. Thank you!

brettsam commented on src/Microsoft.Azure.WebJobs.Host/Hosting/WebJobsBuilderExtensions.cs:
```diff
@@ -58,22 +61,72 @@ public static IWebJobsBuilder UseHostId(this IWebJobsBuilder builder, string hos
+                    logger.LogInformation("Service type: {serviceName}, Implementation: {implementation}, Registered with lifetime: {lifetime}", serviceName, implementation, lifetime);
+                }
+
+                tracker.ResetTracking();
```
Do we need to do this here? If not, can we remove this API from the interface?

fabiocav replied:
You need to be able to reset tracking with every iteration (every startup), correct?

madelinegordon replied:
Yes, tracking should be reset so that only newly added services on startup are logged.

brettsam replied:
Ahh, right.

brettsam reviewed (commented):
Nice! I have a bunch of nitpicky comments -- let me know if you have questions on any of them.

fabiocav commented on src/Microsoft.Azure.WebJobs.Host/Hosting/ServiceCollectionDecorator.cs:
```diff
@@ -0,0 +1,98 @@
+    {
+        public IServiceCollection ServiceCollection { get; set; }
+
+        public List<ServiceDescriptor> Diff { get; private set; } // tracks any service collection modification.
```
Same comment on naming (something like `TrackedChanges` would be appropriate here)

fabiocav commented on src/Microsoft.Azure.WebJobs.Host/Hosting/ServiceCollectionDecorator.cs:
```diff
@@ -0,0 +1,98 @@
+
+        public ServiceCollectionDecorator(IServiceCollection serviceCollection)
+        {
+            ServiceCollection = serviceCollection;
```
perform a `null` check here.

fabiocav commented on src/Microsoft.Azure.WebJobs.Host/Hosting/WebJobsBuilder.cs:
```diff
@@ -12,9 +13,10 @@ internal class WebJobsBuilder : IWebJobsBuilder
         }
 
-        public IServiceCollection Services { get; }
+        public IServiceCollection Services { get; }     
```
Empty trailing spaces

fabiocav commented on src/Microsoft.Azure.WebJobs.Host/Hosting/WebJobsBuilderExtensions.cs:
```diff
@@ -58,22 +61,72 @@ public static IWebJobsBuilder UseHostId(this IWebJobsBuilder builder, string hos
+        {
+            ITrackingServiceCollection tracker = builder.Services as ITrackingServiceCollection;
+
+            if (tracker != null)
```
Change to:
```
if (builder.Services is ITrackingServices tracker)
```

fabiocav commented on src/Microsoft.Azure.WebJobs.Host/Hosting/WebJobsBuilderExtensions.cs:
```diff
@@ -58,22 +61,72 @@ public static IWebJobsBuilder UseHostId(this IWebJobsBuilder builder, string hos
+                    string implementation = "";
+                    if (service.ImplementationInstance != null)
+                    {
+                        implementation = service.ImplementationInstance.GetType().FullName + "[instance]";
```
nit: add a space before ` [instance]`

fabiocav commented on src/Microsoft.Azure.WebJobs.Host/Hosting/WebJobsBuilderExtensions.cs:
```diff
@@ -58,22 +61,72 @@ public static IWebJobsBuilder UseHostId(this IWebJobsBuilder builder, string hos
+                    {
+                        implementation = service.ImplementationInstance.GetType().FullName + "[instance]";
+                    }
+                    else if (service.ImplementationType != null)
```
check implementation type first (more common case)

madelinegordon commented:
@brettsam @fabiocav I have updated the PR and addressed all comments that were there so far--let me know if there is anything else you find to be updated. Otherwise, will just need your approval (although it looks like there isn't a policy requiring an official approval in this repo). Thanks!

brettsam commented on src/Microsoft.Azure.WebJobs.Host/Hosting/ServiceCollectionDecorator.cs:
```diff
@@ -0,0 +1,100 @@
+                return _trackedChanges as IEnumerable<ServiceDescriptor>;
+            }
+            private set { }
+        }
```
For this you should be able to just do (it just makes it shorter).

` public IEnumerable<ServiceDescriptor> TrackedCollectionChanges => _trackedChanges;`

brettsam commented on src/Microsoft.Azure.WebJobs.Host/Hosting/ServiceCollectionDecorator.cs:
```diff
@@ -0,0 +1,100 @@
+            private set { }
+        }
+
+        private List<ServiceDescriptor> _trackedChanges { get; set; }
```
nit: move private fields up to the top of the class.

brettsam commented on src/Microsoft.Azure.WebJobs.Host/Hosting/WebJobsBuilderExtensions.cs:
```diff
@@ -58,22 +62,77 @@ public static IWebJobsBuilder UseHostId(this IWebJobsBuilder builder, string hos
+                    foreach (ServiceDescriptor service in tracker.TrackedCollectionChanges)
+                    {
+                        string serviceName = "Service name: " + service.ServiceType.FullName;
+                        StringBuilder registeredService = new StringBuilder(", Implementation: ");
```
Some logging string nits :-)

1. Don't create a new `StringBuilder` here -- use the existing one `sb` from above.
2. Rather than storing a separate string here for `serviceName` -- just append it. It'll make it a little easier to follow. (see below)
3. Saying "Service name" on every line isn't required as we'll understand it -- and it's mostly for debugging so we can be a little short with it. (see below)
4. Rather than doing "[instance]" -- just replace "Implementation" with "Instance" on those lines. (see below)
5. Use `sb.AppendLine()` -- which I think will use the right line feed mechanics, depending on Linux/Windows. Similarly, I'd use explicit spaces rather than `\t`.

So you can start each line with with:
`sb.AppendLine($"  {service.ServiceType.FullName}: ");` (note just 2 spaces here)

Then depending on whether Implemention/Instance/Factory:
`sb.Append($"Implementation: {service.ImplementationType.FullName}");`
...
`sb.Append($"Instance: {service.ImplementationInstance.GetType().FullName}");`
...
`sb.Append("Factory");`

madelinegordon replied:
I think AppendLine adds a newline to the end of the string, so could use with the first and last strings. Otherwise, I have addressed the rest of the comments. Thanks Brett!

brettsam commented on src/Microsoft.Azure.WebJobs.Host/Hosting/WebJobsBuilderExtensions.cs:
```diff
@@ -58,22 +62,77 @@ public static IWebJobsBuilder UseHostId(this IWebJobsBuilder builder, string hos
+                        string lifetime = ", Lifetime: " + service.Lifetime.ToString();
+                        sb.Append("\n\t" + serviceName + registeredService.ToString() + lifetime);
+                    }
+                    logger.LogDebug(sb.ToString(), new EventId(500, "ExternalStartupServices"));
```
`EventId` needs to be the first parameter here. -- there's an overload that allows for `object` to be passed, which is why this is working, but the event wouldn't get picked up by kusto.

brettsam commented on src/Microsoft.Azure.WebJobs.Host/Hosting/ServiceCollectionDecorator.cs:
```diff
@@ -0,0 +1,100 @@
+
+namespace Microsoft.Azure.WebJobs.Host.Hosting
+{
+    internal class ServiceCollectionDecorator : IServiceCollection, ITrackedServiceCollection
```
Another tiny nit -- maybe call this `TrackedServiceCollection`? We typically don't use `Decorator`, even if the class is a decorator. Better to describe what it's being used for.

fabiocav replied:
I had that feedback, with the exact name, in a previous commend :)

brettsam reviewed (commented):
Some naming/string nits -- functionally looks good!

brettsam commented on src/Microsoft.Azure.WebJobs.Host/Hosting/WebJobsBuilderExtensions.cs:
```diff
@@ -58,22 +61,76 @@ public static IWebJobsBuilder UseHostId(this IWebJobsBuilder builder, string hos
+
+                    foreach (ServiceDescriptor service in tracker.TrackedCollectionChanges)
+                    {
+                        sb.Append($"\n {service.ServiceType.FullName}: ");
```
Does this appear properly formatted? Should it be "\r\n"? https://docs.microsoft.com/en-us/dotnet/api/system.environment.newline?view=netframework-4.8#property-value.

madelinegordon replied:
Yes, you are right! I will create a bug fix.
